src/fairchem/core/modules/element_references.py

```diff
@@ -0,0 +1,206 @@
+
+    from torch_geometric.data import Batch
+
+
+class LinearReference(nn.Module):
+    """Compute a linear reference for target scalar properties"""
+
+    def __init__(
+        self,
+        element_references: torch.Tensor | None = None,
+        max_num_elements: int = 118,
+    ):
+        """
+        Args:
+            element_references (Tensor): tensor with linear reference values
+            max_num_elements (int): max number of elements - 118 is a stretch
+        """
+        super().__init__()
+        self.register_buffer(
+            name="element_references",
+            tensor=element_references
+            if element_references is not None
+            else torch.zeros(max_num_elements + 1),
+        )
+
+    def _apply_refs(
+        self, target: torch.Tensor, batch: Batch, sign: int, reshaped: bool = True
+    ) -> torch.Tensor:
+        """Apply references batch-wise"""
+        indices = batch.atomic_numbers.to(
+            dtype=torch.int, device=self.element_references.device
+        )
+        elemrefs = sign * self.element_references[indices].to(dtype=target.dtype)
+        # this option should not exist, all tensors should have compatible shapes in dataset and trainer outputs
+        if reshaped:
+            elemrefs = elemrefs.view(batch.natoms.sum(), -1)
+
+        return target.index_add(0, batch.batch, elemrefs)
+
+    @torch.autocast(device_type="cuda", enabled=False)
+    def dereference(
+        self, target: torch.Tensor, batch: Batch, reshaped: bool = True
+    ) -> torch.Tensor:
+        """Remove linear references"""
+        return self._apply_refs(target, batch, -1, reshaped=reshaped)
+
+    @torch.autocast(device_type="cuda", enabled=False)
+    def forward(
+        self, target: torch.Tensor, batch: Batch, reshaped: bool = True
+    ) -> torch.Tensor:
+        """Add linear references"""
+        return self._apply_refs(target, batch, 1, reshaped=reshaped)
+
+
+def create_element_references(
+    file: str | Path | None = None,
+    state_dict: dict | None = None,
+) -> LinearReference:
+    """Create an element reference module.
+
+    Currently only linear references are supported.
+
+    Args:
+        type (str): type of reference (only linear implemented)
+        file (str or Path): path to pt or npz file
+        state_dict (dict): a state dict of a element reference module
+
+    Returns:
+        LinearReference
+    """
+
+    # path takes priority if given
+    if file is not None:
+        try:
+            # try to load a pt file
+            state_dict = torch.load(file)
+        except RuntimeError:  # try to read an npz file
+            state_dict = {}
+            with np.load(file) as values:
+                # legacy linref files:
+                if "coeff" in values:
+                    state_dict["element_references"] = torch.tensor(values["coeff"])
+                else:
+                    state_dict["element_references"] = torch.tensor(
```

Comment:
This might be answered somewhere, else but just wondering on which device this ends up on, assuming CPU?

Reply:
Good question! Since the fits are done on CPU, I move them over to the trainers device here:
https://github.com/FAIR-Chem/fairchem/blob/410f711155b0348cf1f9a8740e03b581b8c1169e/src/fairchem/core/trainers/base_trainer.py#L467-L479